internal/users/manager.go

```diff
@@ -73,71 +121,86 @@ func (m *Manager) Stop() error {
-	}
-	// Keep the old UID if the user already exists in the database, to avoid permission issues with the user's home
-	// directory and other files.
-	if !errors.Is(err, cache.NoDataFoundError{}) {
-		u.UID = oldUser.UID
+		return fmt.Errorf("could not get user %q: %w", u.Name, err)
 	}
+	if errors.Is(err, cache.NoDataFoundError{}) {
+		// The user does not exist, so we register it with a unique UID.
+		uid, cleanup, err := m.registerUser(u.Name, false)
+		if err != nil {
+			return fmt.Errorf("could not register user %q: %w", u.Name, err)
+		}
+
+		defer cleanup()
 
-	// Generate the UID of the user unless a UID is already set.
-	if u.UID == 0 {
-		u.UID = m.GenerateUID(u.Name)
+		u.UID = uid
+		oldGroups = []cache.GroupDB{}
+	} else {
+		u.UID = oldUser.UID
+		oldGroups, err = m.cache.GroupsOfUser(oldUser.UID)
+		if err != nil {
+			return fmt.Errorf("could not get groups of user %q: %v", u.Name, err)
+		}
 	}
 
 	// Prepend the user private group
 	u.Groups = append([]GroupInfo{{Name: u.Name, UGID: u.Name}}, u.Groups...)
 
-	// Generate the GIDs of the user groups
-	for i, g := range u.Groups {
+	var authdGroups []cache.GroupDB
+	var localGroups []string
+	for _, g := range u.Groups {
 		if g.Name == "" {
 			return fmt.Errorf("empty group name for user %q", u.Name)
 		}
 
-		if g.UGID == "" {
-			// An empty UGID means that the group is a local group, so we don't need to store a GID for it.
+		if g.UGID == "" && g.GID == nil {
+			// An empty UGID means that the group is local.
+			localGroups = append(localGroups, g.Name)
 			continue
 		}
 
-		// Check if the group already exists in the database
-		oldGroup, err := m.cache.GroupByName(g.Name)
-		if err != nil && !errors.Is(err, cache.NoDataFoundError{}) {
-			return err
-		}
-		// Keep the old GID if the group already exists in the database, to avoid permission issues
-		if !errors.Is(err, cache.NoDataFoundError{}) {
-			u.Groups[i].GID = &oldGroup.GID
-		}
-
-		// Generate the GID of the group unless a GID is already set.
-		if u.Groups[i].GID == nil || *u.Groups[i].GID == 0 {
-			gidv := m.GenerateGID(u.Groups[i].UGID)
-			u.Groups[i].GID = &gidv
+		// Check if the group exists in the database.
+		for _, oldGroup := range oldGroups {
+			if oldGroup.Name == g.Name {
+				// The group already exists in the database, so we keep the GID.
+				g.GID = &oldGroup.GID
+			}
 		}
-	}
 
-	var groupContents []cache.GroupDB
-	var localGroups []string
-	for _, g := range u.Groups {
-		// Empty GID assume local group
 		if g.GID == nil {
-			localGroups = append(localGroups, g.Name)
-			continue
+			// The group does not exist in the database, so we generate a new GID.
+			gid, cleanup, err := m.registerGroup(g.Name)
+			if err != nil {
+				return fmt.Errorf("could not generate GID for group %q: %v", g.Name, err)
+			}
+
+			defer func() {
+				cleanupErr := cleanup()
+				if cleanupErr != nil {
+					err = errors.Join(err, fmt.Errorf("could not remove temporary group %q: %v", g.Name, cleanupErr))
```

Comment:
Why not use `fmt.Errorf` with `%w` to pass the old `err`?

Reply:
that would wrap the error. we don't want the cleanup error to wrap the non-cleanup error, because if we did `errors.Is` and `errors.As` could produce incorrect results